core/ir/riscv64/codec.c

```diff
@@ -1777,6 +1796,29 @@ encode_u_imm_opnd(instr_t *instr, byte *pc, int idx, uint32_t *out)
     return true;
 }
 
+/* Encode the immediate field of the U-type format (PC-relative):
+ * |31        12|11   7|6      0|
+ * | imm[31:12] |  rd  | opcode |
+ *  ^----------^
+ * From:
+ * |31        12|11  0|
+ * | imm[31:12] |  0  |
+ */
+static bool
+encode_u_immpc_opnd(instr_t *instr, byte *pc, int idx, uint32_t *out)
+{
+    opnd_t opnd = instr_get_src(instr, idx);
+    uint32_t imm;
+    if (opnd.kind == PC_kind)
+        imm = opnd_get_pc(opnd) - pc;
+    else if (opnd.kind == INSTR_kind)
+        imm = (byte *)opnd_get_instr(opnd)->offset - (byte *)instr->offset;
+    else
+        return false;
+    *out |= SET_FIELD(imm >> 12, 31, 12);
```

Comment:
> Seems to me that the set of acceptable offsets are only the ones with nothing in their lower 12 bits.

Yes, if the lower 12 bits are non-zero, they will be thrown away.

Reply:
I mean to ask: should we throw an error if the bottom 12 bits are non-zero. Is that better than encoding an instr which the user did not intend to.

Reply:
Yeah, I think doing an assertion here is better, and the corresponding test also needs a rewrite.

Reply:
If we do an assertion here, `instr_is_encoding_possible()` will stop working for auipc. :(

```
bool
encoding_possible(decode_info_t *di, instr_t *in, const instr_info_t *ii)
{
    uint enc;

    byte tmp[RISCV64_INSTR_SIZE];
    enc = encode_common(&tmp[0], in, di);
    return enc != ENCFAIL;
}
```

Reply:
Shouldn't encoding_possible also simply fail for OP_auipc if the provided offset has a non-zero bit in its last 12 bits?

Reply:
Oh, the instr doesn't have the offset as an operand anymore, it has the target pc. The offset is computed as the target_pc - encode_pc later.

@derekbruening the RISC-V auipc instr is a bit complex to encode. It should fail when the `offset & 0xfff != 0`. So it depends on the pc where it is being encoded. But encoding_possible doesn't take in the encode pc.

I guess there are rip-relative instrs on x86 that depend on the encode pc, but encoding_possible doesn't need to know the encode pc to decide whether or not it is encode-able, unlike here.

Reply:
I believe that other arches like x86 do *not* check for reachability in encoding_possible and instead fail later in the full encode; I think that was not considered ideal but that is the current state.  I could see it either way: living with that behavior here, or trying to clean it up and passing an encode pc: but it may be complex if not every caller has an encode pc (I'm pretty sure some callers do not...like instr_zeroes_zmmh()?)

Reply:
So we could set `decode_info_t.check_reachable` to false in get_encoding_info (like we do for AARCHXX), and use it to skip the correctness check for the provided target pc.

Reply:
> So we could set decode_info_t.check_reachable to false.

Currently, in `encode_u_immpc_opnd()`, we don't have `decode_info_t* di` available. So it requires a lot of changes in the encoder to pass it down. Should I do this in this PR or submit a new one?

```
static bool
encode_u_immpc_opnd(instr_t *instr, byte *pc, int idx, uint32_t *out);
````

Reply:
Sure. Let's create a new issue with the required details so someone can pick it up in future, and add a TODO here. I assume encode_u_immpc_opnd will simply drop the last 12 bits for now?

Reply:
Added a FIXME in https://github.com/DynamoRIO/dynamorio/pull/6208/commits/7161172061eb98915096a7ba53cc03bed5e0a901.